microbench/src/main/java/io/netty/handler/codec/Base64Benchmark.java

```diff
@@ -0,0 +1,108 @@
+import io.netty.handler.codec.base64.Base64Dialect;
+import io.netty.microbench.util.AbstractMicrobenchmark;
+import io.netty.util.ResourceLeakDetector;
+import io.netty.util.internal.ThreadLocalRandom;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+import org.openjdk.jmh.annotations.Warmup;
+
+import java.nio.ByteBuffer;
+
+@State(Scope.Benchmark)
+@Warmup(iterations = 10)
+@Measurement(iterations = 20)
+public class Base64Benchmark extends AbstractMicrobenchmark {
+    private final ResourceLeakDetector.Level oldLevel = ResourceLeakDetector.getLevel();
+    private ByteBufAllocator heapAllocator;
+    private ByteBufAllocator directAllocator;
+    private ByteBufAllocator pooledDirectAllocator;
+
+    @Param({ "1024", "65536" })
+    public int size;
+
+    @Param({ "DISABLED", "SIMPLE", "ADVANCED", "PARANOID" })
+    public String level;
+
+    private ByteBuf src;
+
+    @Setup
+    public void setup() {
+        ResourceLeakDetector.setLevel(ResourceLeakDetector.Level.valueOf(level));
+
+        byte[] bytes = new byte[size];
+        ThreadLocalRandom.current().nextBytes(bytes);
+        src = Unpooled.wrappedBuffer(bytes);
+
+        heapAllocator = new UnpooledByteBufAllocator(false);
+        directAllocator = new UnpooledByteBufAllocator(true);
+        pooledDirectAllocator = new PooledByteBufAllocator(true);
+    }
+
+    @TearDown
+    public void tearDown() {
+        src.release();
+        ResourceLeakDetector.setLevel(oldLevel);
+    }
+
+    @Benchmark
+    public boolean javaBase64Encode() {
+        return toBase64Java8(src).release();
+    }
+
+    @Benchmark
+    public boolean nettyBase64EncodeDirectDst() {
+        return encode(directAllocator, src).release();
+    }
+
+    @Benchmark
+    public boolean nettyBase64EncodePooledDirectDst() {
+        return encode(pooledDirectAllocator, src).release();
+    }
+
+    @Benchmark
+    public boolean nettyBase64EncodeHeapDst() {
+        return encode(heapAllocator, src).release();
+    }
+
+    private static ByteBuf encode(ByteBufAllocator allocator, ByteBuf src) {
+        ByteBuf dst = Base64.encode(src, src.readerIndex(),
+                src.readableBytes(), true, Base64Dialect.STANDARD, allocator);
+        src.readerIndex(src.writerIndex());
+        return dst;
+    }
+
+    private static ByteBuf toBase64Java8(ByteBuf src) {
+        ByteBuffer srcBuffer = src.internalNioBuffer(src.readerIndex(), src.readableBytes());
+        int remaining = srcBuffer.remaining();
+        ByteBuffer encodedBuffer = java.util.Base64.getMimeEncoder(76, new byte[]{ '\n' }).encode(srcBuffer);
+        src.skipBytes(remaining - srcBuffer.remaining());
+        return Unpooled.wrappedBuffer(encodedBuffer);
```

Comment:
can we also measure the impacts of going back to direct buffer ... the JDK will always produce heap based buffers and we need them to be direct.